app/queries/keyword_search_entries_query.rb

```diff
@@ -0,0 +1,36 @@
+# frozen_string_literal: true
+
+class KeywordSearchEntriesQuery
+  def initialize(keyword_id:, relation: KeywordSearchEntry)
+    relation = relation.where(keyword_id: keyword_id)
+    @relation = relation
+  end
+
+  def top_ads_count
+    relation.where(kind: :ads, position: :top).size
+  end
+
+  def total_ads_count
+    relation.where(kind: :ads).size
+  end
+
+  def non_ads_count
+    relation.where(kind: :non_ads).size
+  end
+
+  def top_ads_urls
+    urls_of(relation.where(kind: :ads, position: :top))
+  end
+
+  def non_ads_urls
+    urls_of(relation.where(kind: :non_ads))
```

Comment:
here we call the `relation.where(kind: :non_ads)` 2 times from 2 different methods, we know the Active record will cache the query result on their side, but to make the code more clear, and also caching it on the Ruby side by using [||= operator](https://stackoverflow.com/questions/995593/what-does-or-equals-mean-in-ruby), should we refactor it like

![image](https://github.com/Goose97/google-scraper-ruby/assets/11751745/12bca325-5534-47d9-9d0e-18d8e8c3eb76)

```
def non_ads_count
   non_ads.size
 end

def non_ads_urls
  urls_of(non_ads)
end

private

def non_ads
  @non_ads ||= relation.where(kind: :non_ads)
end

# Same for top_ads
def top_ads
  @top_ads ||= relation.where(kind: :ads, position: :top)
end
```

Reply:
Nice suggestion. Added in 45cf7722527dfd3fa1f64acb540c5fb57ef43355